xformers/benchmarks/benchmark_mem_eff_attention.py

```diff
@@ -5,172 +5,236 @@
+        attn_bias_type,
+        batch_size=B,
+        q_len=M,
+        kv_len=M,
+        device=device,
+        dtype=dtype,
+    )
+
+    dtype_str = {
+        torch.bfloat16: "b16",
+        torch.half: "f16",
+        torch.float: "f32",
+    }[dtype]
+    sub_label = f"{dtype_str} {op.NAME} B={B}, M={M}, K={K}"
+
+    if True:
+        r = xformers.ops.memory_efficient_attention(q, q, q, attn_bias, op=op).float()
+        rr = ref_attention(
+            q.float(),
+            q.float(),
+            q.float(),
+            attn_bias,
+        )
+        assert (r - rr).abs().max() < 4e-3, (r - rr).abs().max()
+        del r, rr
+
+    yield benchmark.Timer(
+        stmt="fn(q, q, q, attn_bias, p)",
+        globals={
+            "q": q,
+            "attn_bias": attn_bias,
+            "p": p,
+            "fn": partial(xformers.ops.memory_efficient_attention, op=op),
+        },
+        label=f"attention (attn_bias={attn_bias_type})",
+        description="optimized",
+        sub_label=sub_label,
+        num_threads=num_threads,
+    )
+    yield benchmark.Timer(
+        stmt="fn(q, q, q, attn_bias, p)",
+        globals={
+            "q": q,
+            "attn_bias": attn_bias,
+            "p": p,
+            "fn": ref_attention,
+        },
+        label=f"attention (attn_bias={attn_bias_type})",
+        description="vanilla",
+        sub_label=sub_label,
+        num_threads=num_threads,
+    )
+
+
+def benchmark_backward(shape, num_threads: int, attn_bias_type, dtype):
+    B, M, K = shape
+    q = torch.rand(shape, device=device, dtype=dtype, requires_grad=True)
+
+    dispatch = xformers.ops.AttentionOpDispatch(
+        dtype=dtype,
+        device=device,
+        k=K,
+        attn_bias_type=attn_bias_type,
+        has_dropout=False,
+        kv_len=M,
+        q_len=M,
+    )
+    try:
+        op = dispatch.op if FORCE_OP is None else FORCE_OP
+    except NotImplementedError:
+        return
+    if not op.supports(dispatch):
+        return
+
+    attn_bias = create_attn_bias(
+        attn_bias_type,
+        batch_size=B,
+        q_len=M,
+        kv_len=M,
+        device=device,
+        dtype=dtype,
+    )
+
+    dtype_str = {
+        torch.bfloat16: "b16",
+        torch.half: "f16",
+        torch.float: "f32",
+    }[dtype]
+    sub_label = f"{dtype_str} {op.NAME} B={B}, M={M}, K={K}"
+
+    if True:
+        r = xformers.ops.memory_efficient_attention(q, q, q, attn_bias, op=op)
+        r.backward(torch.ones_like(q))
+
+        grad = cast(torch.Tensor, q.grad)
+        q.grad = None
+
+        rr = ref_attention(q, q, q, attn_bias)
+        rr.backward(torch.ones_like(q))
+        atol = 2e-4 + 2e-6 * K * M * math.sqrt(B) * math.sqrt(M)
+        # type: ignore
+        assert (grad - q.grad).abs().max() < atol, f"{(grad - q.grad).abs().max()}"
+        q.grad = None
+        del r, rr, grad
+
+    out = xformers.ops.memory_efficient_attention(q, q, q, attn_bias, p, op=op)
```

Comment:
same as above (self vs. non self), maybe that the numbers end up being similar but I would be curious to see that for real. Guess is that you would slightly increase the gap vs. pytorch for non self-attention